controller-gen panics when providing schema with nested structs + struct tags

This issue seems to be similar to Issue #442, but i am not sure if the root cause is the same. 

When writing a Schema, containing a nested struct, the `controller-gen` tool panics. 

```go
type MySpec struct {
	TestStruct struct {
		Test1 string `json:"test1,omitempty"`
	} `json:"test_struct"`
}
```

will lead to 
```bash
➜  memcached-operator make manifests
/Users/wzhjeki/go/src/code.rbi.tech/WZHJEKI/operatorsdktests/memcached-operator/bin/controller-gen rbac:roleName=manager-role crd webhook paths="./..." output:crd:artifacts:config=config/crd/bases
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1613e48]

goroutine 1 [running]:
sigs.k8s.io/controller-tools/pkg/crd.structToSchema(0xc001a4e2d0, 0xc000e5b7e8)
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:336 +0xa8
sigs.k8s.io/controller-tools/pkg/crd.typeToSchema(0xc001a4e2d0, {0x18e91b0, 0xc000e5b7e8})
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:178 +0x145
sigs.k8s.io/controller-tools/pkg/crd.structToSchema(0xc00145a0f0, 0xc000e5b800)
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:393 +0x77a
sigs.k8s.io/controller-tools/pkg/crd.typeToSchema(0xc00145a0f0, {0x18e91b0, 0xc000e5b800})
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:178 +0x145
sigs.k8s.io/controller-tools/pkg/crd.infoToSchema(0xc00145a0f0)
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:121 +0xed
sigs.k8s.io/controller-tools/pkg/crd.(*Parser).NeedSchemaFor(0xc0005424e0, {0xc0001558e0, {0xc000ed98f0, 0x0}})
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/parser.go:190 +0x2ee
sigs.k8s.io/controller-tools/pkg/crd.(*schemaContext).requestSchema(0xc0004ca0f0, {0x0, 0x17e62f6}, {0xc000ed98f0, 0x0})
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:104 +0xd5
sigs.k8s.io/controller-tools/pkg/crd.localNamedToSchema(0xc001a4e1b0, 0xc001250020)
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:234 +0x216
sigs.k8s.io/controller-tools/pkg/crd.typeToSchema(0xc001a4e1b0, {0x18e8e80, 0xc001250020})
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:168 +0x7c
sigs.k8s.io/controller-tools/pkg/crd.structToSchema(0xc00145aae8, 0xc000e5b950)
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:393 +0x77a
sigs.k8s.io/controller-tools/pkg/crd.typeToSchema(0xc00145aae8, {0x18e91b0, 0xc000e5b950})
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:178 +0x145
sigs.k8s.io/controller-tools/pkg/crd.infoToSchema(0xc000216ae8)
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:121 +0xed
sigs.k8s.io/controller-tools/pkg/crd.(*Parser).NeedSchemaFor(0xc0005424e0, {0xc0001558e0, {0xc000ed9970, 0x9}})
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/parser.go:190 +0x2ee
sigs.k8s.io/controller-tools/pkg/crd.(*Parser).NeedFlattenedSchemaFor(0xc0005424e0, {0xc0001558e0, {0xc000ed9970, 0x17e4833}})
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/parser.go:202 +0xb7
sigs.k8s.io/controller-tools/pkg/crd.(*Parser).NeedCRDFor(0xc0005424e0, {{0xc00032c02d, 0x0}, {0xc000ed9970, 0x0}}, 0x0)
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/spec.go:85 +0x56e
sigs.k8s.io/controller-tools/pkg/crd.Generator.Generate({0x0, 0x0, {0x0, 0x0, 0x0}, 0x0}, 0xc0000af5e0)
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/gen.go:118 +0x2da
sigs.k8s.io/controller-tools/pkg/genall.(*Runtime).Run(0xc0004ac900)
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/genall/genall.go:213 +0x246
main.main.func1(0xc0002c4500, {0xc000145f40, 0x5, 0x5})
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/cmd/controller-gen/main.go:176 +0x6a
github.com/spf13/cobra.(*Command).execute(0xc0002c4500, {0xc000132130, 0x5, 0x5})
        /Users/wzhjeki/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:856 +0x60e
github.com/spf13/cobra.(*Command).ExecuteC(0xc0002c4500)
        /Users/wzhjeki/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:974 +0x3bc
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/wzhjeki/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:902
main.main()
        /Users/wzhjeki/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/cmd/controller-gen/main.go:200 +0x2c6
make: *** [manifests] Error 2
```

Specifically, when adding the `json:"test_struct"` tag, the panic is triggered. Leaving it out, leads to a proper error: `encountered struct field "TestStruct" without JSON tag in type "MySpec"`. 

## Workaround 
a workaround is possible, by specifying the same struct in an explicit way: 

```go
type TestStruct struct {
	Test1 string `json:"test1,omitempty"`
}

type MemcachedSpec struct {
	TestStruct TestStruct `json:"test_struct"`
}
```

## My environment 
`operator-sdk version: "v1.16.0", commit: "560044140c4f3d88677e4ef2872931f5bb97f255", kubernetes version: "v1.21", go version: "go1.17.5", GOOS: "darwin", GOARCH: "amd64"`
controller-gen version: `sigs.k8s.io/controller-tools/cmd/controller-gen@v0.7.0`. I could reproduce the same error with 0.8.0

--- a/pkg/crd/parser_integration_test.go
+++ b/pkg/crd/parser_integration_test.go
@@ -210,6 +210,20 @@ var _ = Describe("CRD Generation From Parsing to CustomResourceDefinition", func
 			})
 		})
 
+		Context("Inline struct error", func() {
+			BeforeEach(func() {
+				pkgPaths = []string{"./inline_struct_error/..."}
+				expPkgLen = 1
+			})
+			It("should reject inline struct literals with a clear error message", func() {
+				groupKind := schema.GroupKind{Kind: "InlineStruct", Group: "testdata.kubebuilder.io"}
+				parser.NeedCRDFor(groupKind, nil)
+
+				expectedErr := "inline struct types are not supported, use a named type instead"
+				Expect(packageErrors(pkgs[0])).To(MatchError(ContainSubstring(expectedErr)))
+			})
+		})
+
 		Context("OneOf API with invalid marker", func() {
 			BeforeEach(func() {
 				pkgPaths = []string{"./oneof_error/..."}

--- a/pkg/crd/schema.go
+++ b/pkg/crd/schema.go
@@ -443,6 +443,10 @@ func structToSchema(ctx *schemaContext, structType *ast.StructType) *apiextensio
 		Properties: make(map[string]apiextensionsv1.JSONSchemaProps),
 	}
 
+	if ctx.info.RawSpec == nil {
+		ctx.pkg.AddError(loader.ErrFromNode(fmt.Errorf("inline struct types are not supported, use a named type instead"), structType))
+		return props
+	}
 	if ctx.info.RawSpec.Type != structType {
 		ctx.pkg.AddError(loader.ErrFromNode(fmt.Errorf("encountered non-top-level struct (possibly embedded), those aren't allowed"), structType))
 		return props

--- a/pkg/crd/testdata/inline_struct_error/types.go
+++ b/pkg/crd/testdata/inline_struct_error/types.go
@@ -1,0 +1,41 @@
+/*
+Copyright 2026 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +groupName=testdata.kubebuilder.io
+package inlinestructerror
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+
+// InlineStruct tests that inline struct literals are properly rejected.
+// This should fail with a clear error message instead of panicking.
+type InlineStruct struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec InlineStructSpec `json:"spec,omitempty"`
+}
+
+// InlineStructSpec contains an inline struct field which should be rejected.
+type InlineStructSpec struct {
+	// TestStruct is an inline struct literal which is not supported.
+	TestStruct struct {
+		Test1 string `json:"test1,omitempty"`
+	} `json:"test_struct"`
+}

